exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/DrillFunctionRegistry.java

```diff
@@ -218,4 +302,141 @@ private void registerOperatorsWithoutInference(DrillOperatorTable operatorTable)
+
+    // jar name, Map<function name, List<signature>
+    private final Map<T, Map<T, List<T>>> jars;
+
+    // function name, Map<signature, function holder>
+    private final Map<T, Map<T, U>> functions;
+
+    public GenericRegistryHolder() {
+      this.functions = Maps.newHashMap();
+      this.jars = Maps.newHashMap();
+    }
+
+    public void addJar(T jName, Map<T, Pair<T, U>> sNameMap) {
+      try (AutoCloseableLock lock = writeLock.open()) {
+        Map<T, List<T>> map = jars.get(jName);
+        if (map != null) {
+          removeAllByJar(jName);
+        }
+        map = Maps.newHashMap();
+        jars.put(jName, map);
+
+        for (Entry<T, Pair<T, U>> entry : sNameMap.entrySet()) {
+          T sName = entry.getKey();
+          Pair<T, U> pair = entry.getValue();
+          addFunction(jName, pair.getKey(), sName, pair.getValue());
+        }
+      }
+    }
+
+    public void removeJar(T jName) {
+      try (AutoCloseableLock lock = writeLock.open()) {
+        removeAllByJar(jName);
+      }
+    }
+
+    public List<T> getAllJarNames() {
+      try (AutoCloseableLock lock = readLock.open()) {
+        return Lists.newArrayList(jars.keySet());
+      }
+    }
+
+    public List<T> getAllFunctionNames(T jName) {
+      try  (AutoCloseableLock lock = readLock.open()){
+        Map<T, List<T>> map = jars.get(jName);
+        return map == null ? Lists.<T>newArrayList() : Lists.newArrayList(map.keySet());
+      }
+    }
+
+    public ListMultimap<T, U> getAllFunctionsWithHolders() {
+      try (AutoCloseableLock lock = readLock.open()) {
+        ListMultimap<T, U> multimap = ArrayListMultimap.create();
+        for (Entry<T, Map<T, U>> entry : functions.entrySet()) {
+          multimap.putAll(entry.getKey(), Lists.newArrayList(entry.getValue().values()));
+        }
+        return multimap;
+      }
+    }
+
+    public ListMultimap<T, T> getAllFunctionsWithSignatures() {
+      try (AutoCloseableLock lock = readLock.open()) {
+        ListMultimap<T, T> multimap = ArrayListMultimap.create();
+        for (Entry<T, Map<T, U>> entry : functions.entrySet()) {
+          multimap.putAll(entry.getKey(), Lists.newArrayList(entry.getValue().keySet()));
+        }
+        return multimap;
+      }
+    }
+
+    public List<U> getHoldersByFunctionName(T fName) {
+      try (AutoCloseableLock lock = readLock.open()) {
+        Map<T, U> map = functions.get(fName);
+        return map == null ? Lists.<U>newArrayList() : Lists.newArrayList(map.values());
+      }
+    }
+
+    public boolean containsJar(T jName) {
+      try (AutoCloseableLock lock = readLock.open()) {
+        return jars.containsKey(jName);
+      }
+    }
+
+    public int functionsSize() {
+      try (AutoCloseableLock lock = readLock.open()) {
+        return functions.size();
+      }
+    }
+
+    private void addFunction(T jName, T fName, T sName, U fHolder) {
```

Comment:
I can guess that jName is jar name and fName is function name. What is sName? Comments are our fiends, as are descriptive names.